### PR TITLE
feat: check if service is available in the region

### LINKED ISSUE #3172: Check whether AppRunner is available in the region
I ran into a strange error message when trying to deploy a request driven web service in the Frankfurt (eu-central-1) region.

`Template format error: Unrecognized resource types: [AWS::AppRunner::Service]`

After some digging I found out that AppRunner isn't available in this region yet. It would be nice if copilot-cli could check whether a template is valid for the region and issue a warning.

### PR DESCRIPTION
**Pull request by paragbhingre**
This PR adds ability to check if service exist in the given target environment region for every `deploy` command.  

This PR resolves #3172

### CODE REVIEW

**Comment by paragbhingre** [comment]
Oops some test errors, while I check that you can still take a look at implementation 😁

File: internal/pkg/manifest/lb_web_svc.go
Comment (efekarakus) [inline internal/pkg/manifest/lb_web_svc.go]: Can we move the constants to our `internal/aws` pkgs for example:
```go
// in internal/pkg/aws/ecs
package ecs

const ServiceID = ecs.ServiceID
```

We want to do this because it makes it easier to upgrade a dependency (aws-sdk-go) from v1 to v2. This way we have a single location to update `internal/pkg/aws` instead of updating multiple pkgs in the codebase. Hope that makes sense!

File: internal/pkg/aws/regions/regions.go
Comment (efekarakus) [inline internal/pkg/aws/regions/regions.go]: 🥺 https://github.com/golang/go/wiki/CodeReviewComments#package-comments

File: internal/pkg/manifest/lb_web_svc.go
Comment (efekarakus) [inline internal/pkg/manifest/lb_web_svc.go]: 🥺 https://github.com/golang/go/wiki/CodeReviewComments#doc-comments

**Review by efekarakus — COMMENTED** [review]
I haven't looked deeply to the PR, but at a glance it looks great!  
Just some minor feedback, feel free to not wait for my approval if others approve.

File: internal/pkg/manifest/lb_web_svc.go
Comment (efekarakus) [reply internal/pkg/manifest/lb_web_svc.go]: (also I believe it needs to be `.ServiceID` instead of `.EndpointsID`)

Comment (Lou1415926) [reply internal/pkg/manifest/lb_web_svc.go]: I believe `EndpointsID` is used here because of this description in the [aws-sdk-go doc](https://docs.aws.amazon.com/sdk-for-go/api/aws/endpoints/#Endpoint.ServiceID):
> Deprecated: Use client package's EndpointsID value instead of these ServiceIDs. These IDs are not maintained, and are out of date.

Very confusing description since the parameter names are still `ServiceID` 🤔

Comment (Lou1415926) [reply internal/pkg/manifest/lb_web_svc.go]: I went ahead and tested it - 
`apprunner` package's `ServiceID` is `"AppRunner"`, and its `EndpointsID` is `apprunner`. The `ServiceID` when passed into ` endpoints.RegionsForService` doesn't work, while the latter does.

File: internal/pkg/cli/svc_deploy.go
Comment (Lou1415926) [inline internal/pkg/cli/svc_deploy.go]: nit:
```suggestion
		return nil, fmt.Errorf("check if %s is available in region %s: %w ", in.svcType, in.targetEnv.Region, err)
```

File: internal/pkg/cli/svc_deploy.go
Comment (Lou1415926) [inline internal/pkg/cli/svc_deploy.go]: nit:
```suggestion
		log.Warningf(`%s might not be available in region %s, proceed with caution.
```

File: internal/pkg/manifest/backend_svc_test.go
Comment (Lou1415926) [inline internal/pkg/manifest/backend_svc_test.go]:
```suggestion
			if tc.wantErr != nil {
				require.EqualError(t, gotErr, tc.wantErr.Error())
			} else { 
			        require.NoError(t, gotErr)
				require.Equal(t, tc.want, got)
			}
```

File: internal/pkg/manifest/backend_svc_test.go
Comment (Lou1415926) [inline internal/pkg/manifest/backend_svc_test.go]: nit: probably redundant since `wantErr` is initialized as `nil`.
```suggestion
```

File: internal/pkg/aws/regions/regions.go
Comment (Lou1415926) [inline internal/pkg/aws/regions/regions.go]: Can we have unit tests for this function?

Comment (huanjani) [inline internal/pkg/aws/regions/regions.go]: nits:
```suggestion
func IsAvailableInRegion(sID string, region string) (bool, error) {
```
Here and below: can we make the function names more declarations than questions, for consistency, and because they're returning a booleans?

File: internal/pkg/aws/regions/regions.go
Comment (Lou1415926) [inline internal/pkg/aws/regions/regions.go]: This package is dependent on `internal/pkg/aws/partitions` - I think ideally the dependencies of `internal/pkg/aws` should be external package only.  Moreover, `partitions` package also rely on `aws/endpoints` package. So I think it probably makes sense to merge this `regions` package with `partitions` package

File: internal/pkg/manifest/backend_svc_test.go
Comment (Lou1415926) [reply internal/pkg/manifest/backend_svc_test.go]: It's actually not clear what these unit tests in `manifest` package are testing 🤔 These tests seems to be testing `regions.IsServiceAvailable`, instead of `manifest`'s `IsServiceAvailable`

File: internal/pkg/cli/svc_deploy.go
Comment (huanjani) [inline internal/pkg/cli/svc_deploy.go]: nit:
```suggestion
		return nil, fmt.Errorf("check %s availability for region %s: %w ", in.svcType, in.targetEnv.Region, err)
```

**Review by huanjani — COMMENTED** [review]
This feature is going to be so helpful for our users! A couple small comments before I look again more deeply.

File: internal/pkg/cli/svc_deploy.go
Comment (huanjani) [reply internal/pkg/cli/svc_deploy.go]: Haha, I was commenting on this at the same time as @Lou1415926. Her suggestion is better!

File: internal/pkg/cli/svc_deploy.go
Comment (huanjani) [inline internal/pkg/cli/svc_deploy.go]: another tiny nit (for here and elsewhere):
```suggestion
		log.Warningf(`%s might not be available in region %s; proceed with caution
```

File: internal/pkg/manifest/lb_web_svc.go
Comment (paragbhingre) [reply internal/pkg/manifest/lb_web_svc.go]: Yes even I did that testing and ServiceID did not work where as EndpintsID worked, hence used EndpointsID finally.

File: internal/pkg/manifest/backend_svc_test.go
Comment (paragbhingre) [reply internal/pkg/manifest/backend_svc_test.go]: Yeah right, first I thought it is important to have a unit test for each kind of a service for example loadbalanced service or backend service and etc. But then I realized it is enough to write a test for `regions.IsServiceAvailable` with only two scenarios i.e. ecs service and apprunner service. So I have removed tests from other `service_test` files and have only one test for `regions.IsServiceAvailable`
Let me know if that makes sense?

File: internal/pkg/aws/regions/regions.go
Comment (paragbhingre) [reply internal/pkg/aws/regions/regions.go]: Done as per above comment

File: internal/pkg/aws/apprunner/apprunner.go
Comment (Lou1415926) [inline internal/pkg/aws/apprunner/apprunner.go]: What do you think of moving these `EndpointsID` to `aws/partitions` so that all the endpoints are defined in a single package for easier maintenance?

```go
# in aws/partitions

const (
   AppRunnerEndpointsID = apprunner.EndpointsID
   ECSEndpointsID = ecs.EndpointsID
)

# in manifest/backend_svc.go
// IsServiceAvailableInRegion checks if service type exist in the given region.
func (s *BackendService) IsServiceAvailableInRegion(region string) (bool, error) {
	return partitions.IsAvailableInRegion(partitions.ECSEndpointsID, region)
}
```

Comment (iamhopaul123) [inline internal/pkg/aws/apprunner/apprunner.go]: Comment for public var please 🙏

File: internal/pkg/aws/partitions/partition.go
Comment (iamhopaul123) [inline internal/pkg/aws/partitions/partition.go]:
```suggestion
// IsAvailableInRegion returns true if the service ID is available in the given region.
```

File: internal/pkg/aws/partitions/partition_test.go
Comment (Lou1415926) [inline internal/pkg/aws/partitions/partition_test.go]: nit: maybe -
```suggestion
			sID:   ECSEndpointsID,
```

File: internal/pkg/aws/partitions/partition_test.go
Comment (Lou1415926) [inline internal/pkg/aws/partitions/partition_test.go]: nit:
```suggestion
			sID:    AppRunnerEndpointsID,
```

File: internal/pkg/aws/partitions/partition.go
Comment (iamhopaul123) [inline internal/pkg/aws/partitions/partition.go]:
```suggestion
	regionServices, existInPartition := endpoints.RegionsForService(endpoints.DefaultPartitions(), partition.ID(), sID)
	if !existInPartition {
	  return false, nil
	}
	_, existInRegion := regionServices[region]
	return existInRegion, nil
```
In case if the service is not available in the partition.

File: internal/pkg/cli/svc_deploy.go
Comment (iamhopaul123) [inline internal/pkg/cli/svc_deploy.go]: I feel like `IsServiceAvailableInRegion` this should be a method of our [deployer](https://github.com/aws/copilot-cli/blob/183c7302ca142997d618e91bec9e7947b03aacd2/internal/pkg/cli/svc_deploy.go#L115) rather than the manifest struct 🤔

Comment (Lou1415926) [reply internal/pkg/cli/svc_deploy.go]: +1 to this!

File: internal/pkg/cli/svc_package.go
Comment (Lou1415926) [inline internal/pkg/cli/svc_package.go]: If we move `IsServiceAvailableInRegion` to the `deployer` we probably wouldn't need to pass around `svcType` anymore

File: internal/pkg/aws/partitions/partition_test.go
Comment (Lou1415926) [inline internal/pkg/aws/partitions/partition_test.go]: Let's put `us-west-1` for this test because it exists & app runner is not available in this region 👀

**Review by iamhopaul123 — APPROVED** [review]

File: internal/pkg/aws/partitions/partition.go
Comment (Lou1415926) [inline internal/pkg/aws/partitions/partition.go]: tiny tiny nit: after checking the doc I feel like these var names may better descriptions ⬇️ 
```suggestion
	regions, partitionOrServiceExists := endpoints.RegionsForService(endpoints.DefaultPartitions(), partition.ID(), sID)
	if !partitionOrServiceExists {
		return false, nil
	}
	_, existInRegion := regions[region]
```

File: internal/pkg/cli/job_deploy.go
Comment (Lou1415926) [inline internal/pkg/cli/job_deploy.go]: What do you think of checking service availability before `deployer.UploadArtifacts`? This way we error out much quicker.

File: internal/pkg/cli/job_deploy.go
Comment (Lou1415926) [inline internal/pkg/cli/job_deploy.go]:
```suggestion
		log.Warningf(`Scheduled Job might not be available in region %s; proceed with caution.
```
I think it's okay without `(s)` because "Scheduled Job" here refers to a type, not a group of workloads.

File: internal/pkg/cli/svc_deploy.go
Comment (Lou1415926) [inline internal/pkg/cli/svc_deploy.go:200]: Samesiess

**Review by Lou1415926 — APPROVED** [review]
Awesome feature! Feel free to remove the DNM label when you are ready.

**Pull request merged**
